### PR TITLE
Fix socket exception on datacollection in parallel

### PR DESCRIPTION
**Pull request by smadala**
## Description
**Fix**
- Stop the Server/Client on TestExecution.Completed/TestDiscovery.Completed received.
- Make EqtTrace logging friendly to parallel scenario.

## Related issue
https://github.com/Microsoft/vstest/issues/1472

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 1]
```diff
@@ -596,6 +596,7 @@ private void SetOperationComplete()
                 EqtTrace.Verbose("TestRequestSender.SetOperationComplete: Setting operation complete.");
             }
 
+            this.communicationEndpoint.Stop();
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Do we need lock here?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs [hunk 2]
```diff
@@ -25,14 +25,19 @@ internal static class TcpClientExtensions
                 CancellationToken cancellationToken)
         {
             Exception error = null;
+            var remoteEndPoint = ((IPEndPoint)client.Client.RemoteEndPoint).ToString();
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs]: do we need typecast?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs [hunk 3]
```diff
@@ -25,14 +25,19 @@ internal static class TcpClientExtensions
...
             // Set read timeout to avoid blocking receive raw message
             while (channel != null && !cancellationToken.IsCancellationRequested)
             {
+                EqtTrace.Verbose("TcpClientExtensions.MessageLoopAsync: loop starting: remoteendPoint: {0} localEndPoint: {1}", remoteEndPoint, localEndPoint);
+
                 try
                 {
                     if (client.Client.Poll(STREAMREADTIMEOUT, SelectMode.SelectRead))
                     {
+                        EqtTrace.Verbose("TcpClientExtensions.MessageLoopAsync: NotifyDataAvailable remoteendPoint: {0} localEndPoint: {1}", remoteEndPoint, localEndPoint);
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs]: >MessageLoopAsync [](start = 62, length = 16)

Lets see logs for this

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs [hunk 4]
```diff
@@ -175,21 +184,27 @@ public Collection<AttachmentSet> AfterTestRunEnd(bool isCanceled, ITestMessageEv
...
+
+            ChangeConnectionTimeoutIfRequired(dataCollectionProcessId);
+
+            EqtTrace.Info("ProxyDataCollectionManager.Initialize: waiting for connection with timeout: {0}", this.connectionTimeout);
 
-            ChangeConnectionTimeoutIfRequired(processId);
             var connected = this.dataCollectionRequestSender.WaitForRequestHandlerConnection(this.connectionTimeout);
             if (connected == false)
             {
+                EqtTrace.Error("ProxyDataCollectionManager.Initialize: failed to connect to datacollector process.");
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: Add datacollector process ID we could not connect to

**Review by mayankbansal018 — APPROVED** [review]
:shipit:

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs (hunk 1 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Not required.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs (hunk 2 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs (hunk 3 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TcpClientExtensions.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs (hunk 4 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: Done.

**Pull request merged**